### PR TITLE
Fix bug in sort optimization

### PR DESCRIPTION
**Pull request by mayya-sharipova**
Fix bug how iterator with skipping functionality
advances and produces docs

Relates to #1725

### CODE REVIEW

**Comment by mayya-sharipova** [comment]
@jimczi  I could not reproduce an issue of duplicate documents in Lucene, and I am still not sure how it actually happens.

 Even as we were [incorrectly advancing](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/Weight.java#L216) only `scorerIterator`, as `filteredIterator`  is an intersection of `collectorIterator` and `scorerIterator`, a next document from `filteredIterator` must be a document > the current document of `scorerIterator`? No?

There was also a question why we did not detect duplicate documents in the tests. First, they don't happen in the tests, and before I was not using `AssertingIndexSearcher`.  I've changed the tests on _doc sort to use `AssertingIndexSearcher`

**Comment by mayya-sharipova** [comment]
@jimczi I was able to reproduce a bug, and in aaab92a added a test that reproducibly fails for me without this patch.  It is a very particular combination of docs and queries to have a `BitSetIterator` as a `scorerIterator` and  ensure that we go through `DefaultBulkSorter::scoreRange` section for scoring.

File: lucene/core/src/test/org/apache/lucene/search/TestFieldSortOptimizationSkipping.java [hunk 1]
```diff
@@ -432,7 +439,48 @@ public void testDocSortOptimization() throws IOException {
...
+    reader.close();
+    dir.close();
+  }
+
+  /**
+   * Test that sorting on _doc works correctly.
+   * This test goes through DefaultBulkSorter::scoreRange, where scorerIterator is BitSetIterator.
+   * As a conjunction of this BitSetIterator with DocComparator's iterator, we get BitSetConjunctionDISI.
+   * BitSetConjuctionDISI advances based on the DocComparator's iterator, and doesn't consider
+   * that its BitSetIterator may have advanced passed a certain doc.
```
Comment (jimczi) [inline lucene/core/src/test/org/apache/lucene/search/TestFieldSortOptimizationSkipping.java:451]: Should we consider this a bug in `BitSetIterator` ?

**Review by jimczi — APPROVED** [review]
The fix looks good to me. Thanks for adding the test.

File: lucene/core/src/test/org/apache/lucene/search/TestFieldSortOptimizationSkipping.java (hunk 1 above)
Comment (mayya-sharipova) [reply lucene/core/src/test/org/apache/lucene/search/TestFieldSortOptimizationSkipping.java:451]: I will create an issue for this.

Comment (mayya-sharipova) [reply lucene/core/src/test/org/apache/lucene/search/TestFieldSortOptimizationSkipping.java:451]: Issue created: https://issues.apache.org/jira/browse/LUCENE-9541

**Pull request merged**